frame/collective/src/lib.rs

```diff
@@ -168,9 +175,93 @@ decl_error! {
 		AlreadyInitialized,
 		/// The close call is made too early, before the end of the voting.
 		TooEarly,
+		/// An invalid count was provided for weight estimation.
+		InvalidCount,
+		/// There can only be a maximum of `MaxMembers`.
+		TooManyMembers,
+		/// There can only be a maximum of `MaxProposals` active proposals.
+		TooManyProposals,
+	}
+}
+
+/// Functions for calcuating the weight of dispatchables.
+mod weight_for {
+	use frame_support::weights::{GetDispatchInfo, RuntimeDbWeight, Weight};
+
+	/// Calculate the weight for `set_members`.
+	pub(crate) fn set_members(
+		db: RuntimeDbWeight,
+		old_count: impl Into<Weight>,
+		new_count: impl Into<Weight>,
+		proposals: impl Into<Weight> + Copy,
+	) -> Weight {
+		db.reads_writes(1, 1) // read + write members
+			+ db.reads_writes(proposals.into() + 1, proposals.into()) // update votes
+			+ db.writes(2) // set prime
+			+ 29_000_000 * old_count.into() // M
+			+ 100_000 * new_count.into() // N
+			+ 27_000_000 * proposals.into() // P
+	}
+
+	// The bytes are responsible for an insignificant amount of weight.
+	// Length of proposals is limited by length fees.
+	const MAX_ASSUMED_PROPOSAL_BYTES: Weight = 1_024;
+
+	/// Calculate the weight for `execute`.
+	pub(crate) fn execute(
+		db: RuntimeDbWeight,
+		members: impl Into<Weight>,
+		proposal: impl Into<Weight>,
+	) -> Weight {
+		db.reads(1) // read members for `is_member`
+			+ 86_000_000 // constant
+			+ 330_000 * members.into() // M
+			+ 5_000 * MAX_ASSUMED_PROPOSAL_BYTES // B
+			+ proposal.into() // P
+	}
+
+	/// Calculate the weight for `propose` if the proposal is executed straight away (`threshold < 2`).
+	pub(crate) fn propose_execute(
+		db: RuntimeDbWeight,
+		members: impl Into<Weight>,
+		proposal: impl Into<Weight>,
+	) -> Weight {
+		db.reads(2) // `is_member` + `contains_key`
+			+ 120_000_000 // constant
+			+ 650_000 * members.into() // M
+			+ proposal.into() // P1
+			+ 5_000 * MAX_ASSUMED_PROPOSAL_BYTES // B
+		
+	}
+
+	/// Calculate the weight for `propose` if the proposal is put up for a vote (`threshold >= 2`).
+	pub(crate) fn propose_proposed(
+		db: RuntimeDbWeight,
+		members: impl Into<Weight>,
+		proposals: impl Into<Weight>,
+	) -> Weight {
+		db.reads(2) // `is_member` + `contains_key`
+			+ db.reads_writes(2, 4) // `proposal` insertion
+			+ 180_000_000 // constant
+			+ 350_000 * members.into() // M
+			+ 1_100_000 * proposals.into() // P2
+			+ 7_000 * MAX_ASSUMED_PROPOSAL_BYTES // B
 	}
 }
 
+macro_rules! map_result_weight(
+	($result:expr, $closure:expr) => {{
+		match $result {
+			Ok(post_info) => {
+				Ok(post_info.actual_weight.map($closure).into())
+			},
+			Err(err) => {
+				Ok(err.post_info.actual_weight.map($closure).into())
+			}
+		}
+	}}
+);
```

Comment:
What do you guys think about this macro? It feels to me like it increases clarity, but let me know what you think.

Reply:
does it need to be a macro? Or can it just be a function that returns a result?

Reply:
:facepalm: of course